src/library/scala/collection/mutable/HashSet.scala

```diff
@@ -97,10 +261,50 @@ final class HashSet[A]
 @SerialVersionUID(3L)
 object HashSet extends IterableFactory[HashSet] {
 
-  def from[B](it: scala.collection.IterableOnce[B]): HashSet[B] = Growable.from(empty[B], it)
+  def from[B](it: scala.collection.IterableOnce[B]): HashSet[B] = {
+    val k = it.knownSize
+    val cap = if(k > 0) ((k + 1).toDouble / defaultLoadFactor).toInt else defaultInitialCapacity
+    new HashSet[B](cap, defaultLoadFactor) ++= it
+  }
 
   def empty[A]: HashSet[A] = new HashSet[A]
 
-  def newBuilder[A]: Builder[A, HashSet[A]] = new GrowableBuilder(empty[A])
+  def newBuilder[A]: Builder[A, HashSet[A]] = newBuilder(defaultInitialCapacity, defaultLoadFactor)
+
+  def newBuilder[A](initialCapacity: Int, loadFactor: Double): Builder[A, HashSet[A]] =
+    new GrowableBuilder[A, HashSet[A]](new HashSet[A](initialCapacity, loadFactor)) {
+      override def sizeHint(size: Int) = elems.sizeHint(size)
+    }
+
+  /** The default load factor for the hash table */
+  final def defaultLoadFactor: Double = 0.75
+
+  /** The default initial capacity for the hash table */
+  final def defaultInitialCapacity: Int = 16
+
+  @SerialVersionUID(3L)
+  private final class DeserializationFactory[A](val tableLength: Int, val loadFactor: Double) extends Factory[A, HashSet[A]] with Serializable {
+    def fromSpecific(it: IterableOnce[A]): HashSet[A] = new HashSet[A](tableLength, loadFactor) ++= it
+    def newBuilder: Builder[A, HashSet[A]] = HashSet.newBuilder(tableLength, loadFactor)
+  }
+
+  private final class Node[K](
+      @(`inline` @getter) val key: K,
+      @(`inline` @getter) val hash: Int,
+      @(`inline` @getter @setter) var next: Node[K]) {
+
+    @tailrec
+    def findNode(k: K, h: Int): Node[K] =
+      if(h == hash && k == key) this
+      else if((next eq null) || (hash > h)) null
```

Comment:
Curious: is the idea of keeping the linked list sorted by hash new, or was it studied before?

Reply:
I haven't seen it mentioned anywhere but I didn't actively search for it. It's just an idea that popped into my head one night (which made it hard to fall asleep). The results were not as exciting as I hoped for (because bucket collisions are actually quite rare for well-distributed hashes) but still promising.